src/crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.cc

```diff
@@ -235,12 +235,15 @@ LBAInternalNode::find_hole_ret LBAInternalNode::find_hole(
   return seastar::do_with(
     begin,
     L_ADDR_NULL,
-    [this, c, min_addr, len, end=end](auto &i, auto &ret) {
+    [this, c, min_addr, max_addr, len, end=end](auto &i, auto &ret) {
       return trans_intr::repeat([=, &i, &ret]()
         -> find_hole_iertr::future<seastar::stop_iteration> {
-	if (i == end) {
-	  return seastar::make_ready_future<seastar::stop_iteration>(
-	    seastar::stop_iteration::yes);
+        if (i == end) {
+          auto update_end = right_end(max_addr);
```

Comment:
I haven't had time to dig into the bug, but this doesn't look right.  right_end(max_addr) should be the same as end returned from bound().

Reply:
Can you add a test which reproduces the find_hole discrepancy?

Reply:
after bound get [begin, end] iterator, the child node is split. there is a new iterator insert in this node at the end, so the end is changed, and when search split left child node, it use L_ADDR_MAX as the max and got the meta.end laddr in the left child, but this is the right child begin laddr. 

just use fio testing will trigger the bug within 30s,  release build and disable debug output.

Reply:
Yes, but adding a specific unit test will clarify the conditions under which it happens and help use to be sure that this is the right fix.  It'll also prevent reintroducing it in the future.

Reply:
I'm still really not sure about right_end() here.  It seems like under sane conditions it will return the same value as end.

Reply:
I add log print as the following in LBAInternalNode::find_hole: 
 return get_lba_btree_extent(
   c,
   this,
   get_meta().depth - 1,
   i->get_val(),
   get_paddr()
 ).si_then([=, &i](auto extent) mutable {
   auto lb = std::max(min_addr, i->get_key());
   auto ub = i->get_next_key_or_max();
   logger().error("LBAInternalNode::find_hole extent {} lb {} ub {}",
    *extent, lb, ub);
   return extent->find_hole(c, lb, ub, len);
 }).si_then([&i, &ret](auto addr) {
   if (addr == L_ADDR_NULL) {
     ++i;
     return seastar::make_ready_future<seastar::stop_iteration>(
       seastar::stop_iteration::no);
   } else {
            auto it=++i;
     ret = addr;
            logger().error("find hole addr = {} next key = {}", addr, it->get_key());
     return seastar::make_ready_future<seastar::stop_iteration>(
       seastar::stop_iteration::yes);
   }
 });
 }).si_then([&ret, ref=LBANodeRef(this)]() {
   logger().error("ret= {}", ret);
   return ret;
 });

and in LBALeafNode::find_hole :
  logger().error(
    "LBALeafNode::find_hole min={} max={}, end={}, len={}, *this={}",
    min, max, end, len, *this);

I got the following output when add_pin assert happened: 
from the output: ub is  18446744073709551615, and leaf extent end =  4921229312,  ub is get from internal node.   and  find hole addr = 4921229312 next key = 0  ( this output repeat two times, don't known why, but first time next key = 0, the second time next key = 4921229312, just what find hole get addr. seems the internal extent content update at that time. that make me confused also).

and two find hole operation got the same address, this cause the assert happen.  but before the assert happened it is already wrong, since find hole get an existing address. 
----------------------------------
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - LBAInternalNode::find_hole extent CachedExtent(addr=0x600000fc2900, type=LADDR_LEAF, version=7, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 41926656>), paddr=paddr_t<4, 41766912>, state=DIRTY, last_committed_crc=1776592366, refcount=3, size=89, meta=btree_node_meta_t(begin=4901679104, end=4921229312, depth=1)) lb 4901679104 ub 18446744073709551615
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - LBALeafNode::find_hole min=4901679104 max=18446744073709551615, end=4921229312 len=4096, *this=CachedExtent(addr=0x600000fc2900, type=LADDR_LEAF, version=7, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 41926656>), paddr=paddr_t<4, 41766912>, state=DIRTY, last_committed_crc=1776592366, refcount=3, size=89, meta=btree_node_meta_t(begin=4901679104, end=4921229312, depth=1))
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - find hole addr = 4921229312 next key = 0
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - ret= 4921229312
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - find hole addr = 4921229312 next key = 4921229312
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - ret= 4921229312
ERROR 2021-07-02 15:06:44,849 [shard 0] seastore - BtreeLBAManager::alloc_extent: alloc 4921229312~4096 for paddr_t<RECORD_REL_SEG, 16384>, pin laddr 4921229312
........
......
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - LBAInternalNode::find_hole extent CachedExtent(addr=0x600000fc2900, type=LADDR_LEAF, version=7, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 41926656>), paddr=paddr_t<4, 41766912>, state=DIRTY, last_committed_crc=1776592366, refcount=3, size=89, meta=btree_node_meta_t(begin=4901679104, end=4921229312, depth=1)) lb 4901679104 ub 18446744073709551615
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - LBALeafNode::find_hole min=4901679104 max=18446744073709551615, end=4921229312 len=4096, *this=CachedExtent(addr=0x600000fc2900, type=LADDR_LEAF, version=7, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 41926656>), paddr=paddr_t<4, 41766912>, state=DIRTY, last_committed_crc=1776592366, refcount=3, size=89, meta=btree_node_meta_t(begin=4901679104, end=4921229312, depth=1))
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - find hole addr = 4921229312 next key = 0
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - ret= 4921229312
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - find hole addr = 4921229312 next key = 4921229312
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - ret= 4921229312
ERROR 2021-07-02 15:06:44,850 [shard 0] seastore - BtreeLBAManager::alloc_extent: alloc 4921229312~4096 for paddr_t<RECORD_REL_SEG, 20480>, pin laddr 4921229312
ERROR 2021-07-02 15:06:44,851 [shard 0] seastore - add_pin: unable to add btree_range_pin_t(begin=4921229312, end=4921233408, depth=0, extent=0x600000ce0500), found btree_range_pin_t(begin=4921229312, end=4921233408, depth=0, extent=0x600000a27c00)

Reply:
when concurrence increase, for example. I increase iodepth, and numjobs,  I meet two root extent conflict when add pin. 
I add output in root split:
  if (root->at_max_capacity()) {
    split = cache.get_root(t).si_then(
      [this, root, laddr, &t](RootBlockRef croot) {
 logger().error(
   "BtreeLBAManager::insert_mapping: splitting root {}",
   *croot);
 {
          if (!croot->is_pending()) {
     auto mut_croot = cache.duplicate_for_write(t, croot);
     croot = mut_croot->cast<RootBlock>();
          }
 }
 auto nroot = cache.alloc_new_extent<LBAInternalNode>(t, LBA_BLOCK_SIZE);
 logger().error(
   "croot {} ----> nroot {}",
   *croot, *nroot);
 lba_node_meta_t meta{0, L_ADDR_MAX, root->get_node_meta().depth + 1};

the two new allocate root extent conflict in add pin:   I don't know how to solve it now. 

ERROR 2021-07-05 14:17:27,503 [shard 0] seastore - BtreeLBAManager::insert_mapping: splitting root CachedExtent(addr=0x602b980bdf00, type=ROOT, version=4, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 64028672>), paddr=paddr_t<NULL_SEG, NULL_OFF>, state=DIRTY, last_committed_crc=0, refcount=29)
ERROR 2021-07-05 14:17:27,503 [shard 0] seastore - croot CachedExtent(addr=0x600002f56900, type=ROOT, version=5, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 64028672>), paddr=paddr_t<NULL_SEG, NULL_OFF>, state=MUTATION_PENDING, last_committed_crc=0, refcount=3) ----> nroot CachedExtent(addr=0x600000eed980, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<RECORD_REL_SEG, 0>, state=INITIAL_WRITE_PENDING, last_committed_crc=0, refcount=2, size=0, meta=btree_node_meta_t(begin=0, end=0, depth=0))

ERROR 2021-07-05 14:17:27,503 [shard 0] seastore - BtreeLBAManager::insert_mapping: splitting root CachedExtent(addr=0x602b980bdf00, type=ROOT, version=4, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 64028672>), paddr=paddr_t<NULL_SEG, NULL_OFF>, state=DIRTY, last_committed_crc=0, refcount=31)
ERROR 2021-07-05 14:17:27,503 [shard 0] seastore - croot CachedExtent(addr=0x600002f56400, type=ROOT, version=5, dirty_from_or_retired_at=journal_seq_t(segment_seq=4, offset=paddr_t<4, 64028672>), paddr=paddr_t<NULL_SEG, NULL_OFF>, state=MUTATION_PENDING, last_committed_crc=0, refcount=3) ----> nroot CachedExtent(addr=0x60000287fb00, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<RECORD_REL_SEG, 0>, state=INITIAL_WRITE_PENDING, last_committed_crc=0, refcount=2, size=0, meta=btree_node_meta_t(begin=0, end=0, depth=0))

ERROR 2021-07-05 14:17:27,503 [shard 0] seastore - add_pin: unable to add btree_range_pin_t(begin=0, end=18446744073709551615, depth=3, extent=0x600000eed980), found btree_range_pin_t(begin=0, end=18446744073709551615, depth=3, extent=0x60000287fb00)
ERROR 2021-07-05 14:17:27,504 [shard 0] none - ../src/crimson/os/seastore/lba_manager/btree/btree_range_pin.cc:116 : In function 'void crimson::os::seastore::lba_manager::btree::btree_pin_set_t::add_pin(crimson::os::seastore::lba_manager::btree::btree_range_pin_t&)', ceph_assert(%s)
0 == "impossible"

Reply:
I think the root cause is what I mention below, find_hole and insert don't correctly handle changing node boundaries.

Reply:
I still don't understand, why the extent content updated during the continuation? and my log print two times in that continuation.

Reply:
Hmm, I don't have an explanation for that.  Could those be two different transactions?  I think currently we always use laddr=0 as the hint, so two transactions trying to allocate at the same time would probably grab the same address.